Adding well-known configuration and JWKS routes in Laravel

moufmouf opened the pull request:
When using Laravel with Passport, we have enough knowledge of the environment
to be able to provide 2 useful routes our of the box:

- a discovery endpoint at `/.well-known/openid-configuration`.
- a JWKS endpoint at `/oauth/jwks`.

This greatly eases integrating with Laravel since clients can now
use auto-discovery to integrate with Laravel + PassPort + OpenID.

This PR has been successfully tested using the Javascript openid-client package.
This PR is based on top of #13

moufmouf committed:
Adding well-known configuration and JWT routes in Laravel

When using Laravel with Passport, we have enough knowledge of the environment
to be able to provide 2 useful routes our of the box:

- a discovery endpoint at `/.well-known/openid-configuration`.
- a JWKS endpoint at `/oauth/jwks`.

This greatly eases integrating with Laravel since clients can now
use auto-discovery to integrate with Laravel + PassPort + OpenID.

mrmstn commented:
I would be very interested in this PR, is there something left to do @ronvanderheijden? Or is there any case I could help bringing this PR forward?

theaungmyatmoe commented:
I don't think the owner will make this PR forward, will you guys make this own community repo?

moufmouf commented:
@aungmyatmoethegreat I will as a last resort. Before taking such steps, we should ask the repo owner if he needs some help. Maybe he saw the PRs but has no time to deal with those. I opened an issue to ask here: #21

ronvanderheijden commented on src/Laravel/PassportServiceProvider.php:
```diff
@@ -35,26 +35,25 @@ public function boot()
             __DIR__ . '/config/openid.php' => $this->app->configPath('openid.php'),
         ], ['openid', 'openid-config']);
 
+        $this->loadRoutesFrom(__DIR__.'/routes/web.php');
```
I think we should only add the routes when the setting is enable in the config.

moufmouf replied:
Good idea. I added in the config a `routes` entry and one can now select which routes should be enabled (I put the conditions in the `web.php` file)

ronvanderheijden commented on src/Laravel/JwksController.php:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private function getPublicKey(): string {
+        $publicKey = str_replace('\\n', "\n", app()->make(Config::class)->get('passport.public_key') ?? '');
```
```suggestion
        $publicKey = str_replace('\\n', "\n", config('passport.public_key', ''));
```

ronvanderheijden commented on src/Laravel/routes/web.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+Route::get('/oauth/jwks', \OpenIDConnect\Laravel\JwksController::class."@jwks")->name('openid.jwks');
+Route::get('/.well-known/openid-configuration', \OpenIDConnect\Laravel\DiscoveryController::class."@discovery")->name('openid.discovery');
```
```suggestion
use OpenIDConnect\Laravel\JwksController;
use OpenIDConnect\Laravel\DiscoveryController;

Route::get('/oauth/jwks', [JwksController::class, 'jwks'])->name('openid.jwks');
Route::get('/.well-known/openid-configuration', [DiscoveryController::class, 'discovery'])->name('openid.discovery');
```

ronvanderheijden replied:
But an `__invoke()` is perhaps a better improvement.

moufmouf replied:
Let's go for the `__invoke`!

theaungmyatmoe commented on src/Laravel/DiscoveryController.php:
```diff
@@ -0,0 +1,46 @@
+
+class DiscoveryController
+{
+    public function discovery(Request $request)
```
Can't we use an invocable controller while calling a single method?
```suggestion
    public function __invoke(Request $request)
```

moufmouf replied:
Done!

theaungmyatmoe commented on src/IdTokenResponse.php:
```diff
@@ -21,15 +21,18 @@ class IdTokenResponse extends BearerTokenResponse
     private Configuration $config;
+    private ?string $issuer;
 
     public function __construct(
```
I think we can use constructor property promotion in PHP 8.0. So that we can combine class fields, constructor definition and variable assignments all into one syntax, in the construct parameter list.

moufmouf replied:
The package is still compatible with PHP 7.4 (that's what is declared in `composer.json`) so I guess this is a no go for this version at least.

theaungmyatmoe commented on src/Laravel/JwksController.php:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private function getPublicKey(): string {
+        $publicKey = str_replace('\\n', "\n", app()->make(Config::class)->get('passport.public_key') ?? '');
```
In Larvel 11, `config/passport.php` is not published by default, so that we might use passport keypart to retieve the public key.
```suggestion
  $publicKeyFile = File::get(Passport::keyPath('oauth-public.key'));
  $publicKey = str_replace('\\n', "\n", $publicKeyFile);
```

moufmouf replied:
The public key file is not necessarily a file. It can come from env vars for instance. And I believe the `config` method will still read from the config file of teh package (even if the config is not published)

moufmouf committed:
Using the "config" utils instead of making the Config class.

Co-authored-by: Ron van der Heijden <4101322+ronvanderheijden@users.noreply.github.com>

moufmouf committed:
Adding a parameter to make routes configurable.

Routes can now be disabled from the config.
Also, switching to using invokable routes.

moufmouf commented:
@ronvanderheijden Thanks a lot for the review. I took your comments (and those of @aungmyatmoethegreat) in consideration and pushed the change. Tell me if it looks good. I'll rebase the other PRS depending on this one to apply the changes to all depending PRs.

theaungmyatmoe reviewed (approved):
LGTM

moufmouf commented:
Ok, let's merge this! :+1:

ronvanderheijden commented:
This pull-request breaks the actions. I have fixed them in pull-requests #23 :)

moufmouf commented:
Hey @ronvanderheijden,

Really sorry about that. I did not notice the actions. For some reason, they did not show up in the pull request on my side. I'll double check for the other PRs!
